packages/block-library/src/embed/edit.js

```diff
@@ -33,181 +35,215 @@ function getResponsiveHelp( checked ) {
-					title,
-					className,
-					responsive,
-					allowResponsive
-				),
-			};
-		};
-
-		const handleIncomingPreview = () => {
-			setAttributes( getMergedAttributes() );
-			if ( onReplace ) {
-				const upgradedBlock = createUpgradedEmbedBlock(
-					props,
-					getMergedAttributes()
-				);
-
-				if ( upgradedBlock ) {
-					onReplace( upgradedBlock );
-				}
+const EmbedEdit = ( props ) => {
+	const {
+		attributes: {
+			providerNameSlug,
+			previewable,
+			responsive,
+			allowResponsive,
+			caption,
+			url: attributesUrl,
+		},
+		attributes,
+		isSelected,
+		onReplace,
+		setAttributes,
+		insertBlocksAfter,
+	} = props;
+
+	const { icon, title } = getEmbedInfoByProvider( providerNameSlug );
+	const [ url, setURL ] = useState( attributesUrl );
+	const [ isEditingURL, setIsEditingURL ] = useState( false );
+	// const [ mergedAttributes, setMergedAttributes ] = useState( attributes );
+	const { invalidateResolution } = useDispatch( 'core/data' );
+
+	const {
+		preview,
+		fetching,
+		themeSupportsResponsive,
+		cannotEmbed,
+	} = useSelect(
+		( select ) => {
+			const {
+				getEmbedPreview,
+				isPreviewEmbedFallback,
+				isRequestingEmbedPreview,
+				getThemeSupports,
+			} = select( 'core' );
+			if ( attributesUrl === undefined ) {
+				// TODO check if url is also empty... ??
+				return { fetching: false, cannotEmbed: true };
 			}
-		};
 
-		const toggleResponsive = () => {
-			const { allowResponsive, className } = attributes;
-			const { html } = preview;
-			const newAllowResponsive = ! allowResponsive;
-
-			setAttributes( {
-				allowResponsive: newAllowResponsive,
-				className: getClassNames(
-					html,
-					className,
-					responsive && newAllowResponsive
-				),
-			} );
+			const embedPreview = getEmbedPreview( attributesUrl );
+			const previewIsFallback = isPreviewEmbedFallback( attributesUrl );
+
+			// The external oEmbed provider does not exist. We got no type info and no html.
+			const badEmbedProvider =
+				embedPreview?.html === false &&
+				embedPreview?.type === undefined;
+			// Some WordPress URLs that can't be embedded will cause the API to return
+			// a valid JSON response with no HTML and `data.status` set to 404, rather
+			// than generating a fallback response as other embeds do.
+			const wordpressCantEmbed = embedPreview?.data?.status === 404;
+			const validPreview =
+				!! embedPreview && ! badEmbedProvider && ! wordpressCantEmbed;
+			return {
+				preview: validPreview ? embedPreview : undefined,
+				fetching: isRequestingEmbedPreview( attributesUrl ),
+				themeSupportsResponsive: getThemeSupports()[
+					'responsive-embeds'
+				],
+				cannotEmbed: ! validPreview || previewIsFallback,
+			};
+		},
+		[ attributes.url ]
+	);
+
+	/**
+	 * @return {Object} Attributes derived from the preview, merged with the current attributes.
+	 */
+	// TODO maybe use `useCallback` ?? or even better useMemo // or state? with mergedAttributes?
+	const getMergedAttributes = () => {
```

Comment:
// should we use `useCallback` or `useMemo` with dependencies here? Would it better for performance?

Reply:
Considering that, at the source, `getMergedAttributes` is using `memoize`, then it might be nice to:

* Use `useMemo` here as `const mergedAttributes = useMemo( `
* Remove `memoize` from the definition of `getMergedAttributes`.

Reply:
Actually, I take back what I said. There are a few moving parts here and I'd rather approach this carefully. So I would not touch the `memoize` part of `getAttributesFromPreview`.

As for `useMemo` or other approaches... thinking in abstract, it could be nice to replace the way that we use `getMergedAttributes` with a new piece of state (`useState`). The benefit is that it could remove some indirection. The drawback is that it would introduce different life-cycle considerations (any change in that state would cause a re-render, whereas the effect-based dance is more flexible) which could lead to unforeseen bugs.